### ISSUE TITLE
[Enhancement] Streaming API for LOBs

### ISSUE THREAD

**Issue by tgulacsi** [rana/ora#1]
Hi,

I really like the nice simple but thought-out API of your Oracle lib!
It seems that it is a little bit lacking in LOB side.

For example, why is https://github.com/ranaian/ora/blob/master/defLob.go#L98 not a
copy(value[writeIndex:], buffer[:int(byte_amtp)])
?

If you're willing, I can contribute this small change as a PR.

My other question is about LOBs. The can be huge, and as I see you simply read them fully into memory. To mitigate this problem, we could use mmap'd memory for this, or a streaming API (maybe an OraLOBStream type with Read/ReadAt methods?). 
One thing with the streaming API: as I've already burned myself (https://github.com/tgulacsi/goracle/issues/26#issuecomment-73289236), the underlying cursor must be open while reading/writing the LOB, and QueryRow from database/sql closes it after the first row.

Thanks,
Tamás Gulácsi

**Comment by ranaian** [comment]
Hi Tamás,

Thank you for the kind words. The ora package is actually inspired and based on your work in [goracle](https://github.com/tgulacsi/goracle/); so, thank you Tamás. I had a few extra needs to meet so the ora package was built out. LOB wasn't actually one of the needs at the time. Feel free to make any improvements you see fit with LOB or otherwise. It seems you've given thought to LOB optimization and usability. Both 'copy' suggestions and LOB streaming sound great. 

Thank you,

Rana

**Comment by tgulacsi** [comment]
Please help think the API through: technologically we cold add an io.Reader and io.Writer member to Binary struct (in types.go), but I don't know whether the resulting API would be good to use, or not:

```go
type Binary struct {
  IsNull bool
  Value []byte
  io.Reader // will read from here if Value is nil and Reader is not
  io.Writer // will write to here if Value is nil and Writer is not
}
```

Usage:

```go
fh, _ := os.Open("abigfile")
ora.Binary{Reader: fh}

fh, _ = os.Create("destfile")
ora.Binary{Writer:fh}
```

Another solution would be to have separate Binary types: BinaryRead, BinaryWrite, BinaryReadWrite.

**Comment by ranaian** [comment]
Hi Tamás,

Your single struct API suggestion sounds good. It would keep type proliferation to a minimum. Binary already maps to BLOB, RAW and LONG RAW Oracle types. Not adding more types to map might simplify maintenance and learning curve.  

```go
type Binary struct {
	IsNull bool
	Value  []byte
	Reader io.Reader
	Writer io.Writer
}
```

Writing good documentation and examples would be helpful for people to quickly grasp the API differences of `Binary.Reader`, `Binary.Writer`, and `Binary.Value`.

Based on what you said, the API might be:  

1. Preserve `Binary.Value` for single slice usage.  `Binary.Value` can be written as a single slice with the current implementation in [bndBin.go](https://github.com/ranaian/ora/blob/master/bndBin.go) and [bndBinSlice.go](https://github.com/ranaian/ora/blob/master/bndBinSlice.go) using [C.OCI_ONE_PIECE](https://github.com/ranaian/ora/blob/master/bndBin.go#L85). `Binary.Value` can be read as a single slice with the current implementation in [defLob.go](https://github.com/ranaian/ora/blob/master/defLob.go) using [C.OCI_ONE_PIECE](https://github.com/ranaian/ora/blob/master/defLob.go#L109). `Binary.Value` is also used for selecting RAW and LONG RAW columns in [defRaw.go](https://github.com/ranaian/ora/blob/master/defRaw.go) and [defLongRaw.go](https://github.com/ranaian/ora/blob/master/defLongRaw.go). 

2. Add a new `bndBin.bindReader` method which receives a `Binary.Reader` parameter. `bndBin.bindReader` is similar to the existing `bndBin.bind` method using [OCILobWrite2](http://docs.oracle.com/cd/B28359_01/appdev.111/b28395/oci17msc002.htm#i578761). `bndBin.bindReader` would use piece parameters OCI_FIRST_PIECE, OCI_NEXT_PIECE and OCI_LAST_PIECE to read a block from `Binary.Reader` then write a block to Oracle. A similar  `bndBinSlice.bindReader` method may also be created.

3. Modify [defLob.define](https://github.com/ranaian/ora/blob/master/defLob.go#L36) method to use the existing [OCI_ONE_PIECE](https://github.com/ranaian/ora/blob/master/defLob.go#L109) when `Binary.Writer` is nil. When `Binary.Writer` is not nil, add a new call to [OCILobRead2](http://docs.oracle.com/cd/B28359_01/appdev.111/b28395/oci17msc002.htm#i578721) which uses piece parameters OCI_FIRST_PIECE, OCI_NEXT_PIECE and OCI_LAST_PIECE to read a block from Oracle and write a block to `Binary.Writer`.

What you proposed sounds good. Let me know if you have any questions and I will try to help where I can.

Thank you,

Rana

**Comment by tgulacsi** [comment]
Hi,

I've implemented 1-3, but I don't see how to use 3.

See https://github.com/tgulacsi/ora/commits/lob

**Comment by ranaian** [comment]
Wow, amazing turn-around time. 

So, the [defLob.value](https://github.com/ranaian/ora/blob/master/defLob.go#L131) method is called when [*Rset.Next](https://github.com/ranaian/ora/blob/master/rset.go#L166) populates the row columns. ([defLob.value](https://github.com/ranaian/ora/blob/master/defLob.go#L131) is a member of the [def](https://github.com/ranaian/ora/blob/master/type.go#L34) interface.) Within [defLob.value](https://github.com/ranaian/ora/blob/master/defLob.go#L131) you could determine if the user specified a `Binary.Writer`. If `Binary.Writer` is specified,  read from Oracle and write to `Binary.Writer`. I'm not sure if that's the whole of it, but all the `defLob` action occurs in the `value` method.

Let me know if I can clarify anything.

**Comment by ranaian** [comment]
Oh, ok, am thinking this through a little more. Seems that you have thought this though more clearly. The issue with `defLob` is how do we know to write to a `Binary.Value` []byte, or `Binary.Writer` io.Writer.

One option could be to set `Binary.Writer` to a *lobWriter class which would read from Oracle. Methods `Binary.Bytes` and `Binary.String` could be added to optionally allow the user to not have to use `Binary.Writer` if the lob was small enough. 

What do you think?

**Comment by ranaian** [comment]
the more i think about it, the simpler it is to just remove `Binary.Value` and have:

```go
type Binary struct {
    IsNull bool
    Reader io.Reader
    Writer io.Writer
}
```

what do you think?

**Comment by tgulacsi** [comment]
My main problem with defLob.value is _how_ can I decide where to put the LOB's value?
I mean, in which variable/method can I see the user-specified Binary.Writer? Please share a code sample!

**Comment by rana** [comment]
I see your point. Based on your [defLob.go](https://github.com/tgulacsi/ora/blob/d84c8cdcbf3589da4a2eb09dbb3f0f9a71d6e8fc/defLob.go) work, how about providing a `Binary.WriteTo` method?

```go
type Binary struct {
	IsNull bool
	Reader io.Reader  // Reader: for binds, user specifies and driver writes to db
	lr     *lobReader // lr: for defines, driver assigns in defLob.value
}

func (b Binary) WriteTo(w io.Writer) (n int64, err error) {
	if w == nil {
		return "Parameter 'w' is nil."
	}
	return b.lr.WriteTo(w)
}
```

and usage could be

```go
for rset.Next() {
		bin := rset.Row[0].(ora.Binary)
		bin.WriteTo(fileWriter)
	}
```

where the `WriteTo` would have to be called prior to the following call to `rset.Next` when `defLob.free()` would release lob resources. Does this make your life easier? Or is there another approach you would like?

**Comment by tgulacsi** [comment]
Hi,

Finally, a proposal: https://github.com/tgulacsi/ora/tree/lob 
I think some tweaking will be needed, and I haven't thought out the best use cases, and LOBs could be used as files (io.ReaderAt, io.WriterAt), but maybe not needed / unneeded complexity.

So please check it out, and I'm awaiting the critics!

**Comment by rana** [comment]
Hi,

Fantastic work. Thanks also for adding tests. The tests make for easier review and validation. Everything looks good. The only question I would make: are any GoColumnType changes to Bin/OraBin needed to support the Raw and Lob changes? Seems fine as is? Will be nice to see what you decide for reader/writer use cases.

Thank you!

**Comment by tgulacsi** [comment]
GoColumnTypes: maybe.
As Bin/OraBin columns are now sent as LONG RAW, and LOBs retrieved as Lob (not Binary), these are backward-incompatible changes - I'd suggest a major version bump (or go on and move to gopkg.in/rana/ora.v1).

But I'd wait with this, I'm not feeling this lob API ready yet.

**Comment by tgulacsi** [comment]
I'll need some help with the following:
Stmt.bind: do we need *Lob and []*Lob (output bind?)? I've thought that maybe a "SELECT lobCol FROM lobTable FOR UPDATE" would mean open the LOB with READ_WRITE mode - but maybe just use *Lob as an out bind, and treat that as a READ_WRITE mode?

**Comment by rana** [comment]
Generally, Go value types map well to sending data to Oracle during binds and pointer types map well to receiving data from Oracle during binds. I think if that rule is followed consistently across the driver using various types becomes easy to remember which direction data is flowing. From a Go-API consistency point of view, I would make the Lob and []Lob in-bound and *Lob out-bound READ_WRITE. What makes the API easier for you to use? And by extension, easy for everyone else?

**Comment by tgulacsi** [comment]
My other problem is that now we have a Reader.
But LOB's can work as a random access file, too: as a ReaderAt + WriterAt.
But that two modes are incompatible: when a LOB reading is in progress (with OCI_FIRST_PIECE + OCI_NEXT_PIECE), then a ReadAt would need breaking the sequence read, interrupting that.

How should we set the reading/writing mode when defining/binding?
Or is one mode better/adequate/enough?


Sorry for asking so much questions, but the OCI side is ready, and I want as nice & usable extension of your API, as other parts.

**Comment by rana** [comment]
I'm not sure. 
One mode seems enough if random access requires it.
What would you like to see?

**Comment by tgulacsi** [comment]
Ok, the simplest & most useful case is the present:
Lob = struct { io.Reader }, binding a Lob means sending the contents of the Reader, *Lob send + the returned Lob's Reader will read from the LOB. Defining a Lob is just reading from the LOB.

The hard thing is a ReaderAt + WriterAt + Truncate. I've implemented the OCI functions, but...
For bind, Lob could get a Method parameter, which defaults to ReadSeq, but you can set it to RandomAccess, which case we would return a *lobReaderWriterAt (see https://github.com/tgulacsi/ora/commit/ff5a2dbd6554d357f9b42bf790e843fc68328603). 
For define, maybe a new parameter in StmtCfg, with the same values?

Another, more magic solution would be checking whether the SELECT ends with FOR UPDATE, in which case we could do the *lobReaderWriterAt, but I think explicit is better.

If this is not good enough, then I'd just go with the *lobReader side, and postpone the *lobReaderWriterAt, till the need arises, and maybe a better solution.

What do you think?

**Comment by rana** [comment]
Sounds great. Since you've already written `*lobReaderWriterAt` let's add it. Its a fine solution. Adding StmtCfg sounds fine. Regarding adding to StmtCfg or auto checking 'SELECT ends with FOR UPDATE', the most important thing that comes to mind is that there's good usage examples for all the cases you're thinking of in the documentation. Either way you design it, user will understand your intentions.

**Comment by tgulacsi** [comment]
I'm not convinced (by myself :)): I'm always bump into the fact that Lob should have either an io.Reader, or an io.ReaderAt, not both.
If it has both, but one is nil always, then calling a Read may panic, just as calling a ReadAt will panic in the other case.
This suggests for me that we need a separate LobAt type.

TL;DR; I'm not forcing the random access pattern in the proposed PR - maybe later.

**Comment by rana** [comment]
yeah, its a lot of work. If there's a compelling use case then it might be worth it.
